### PR TITLE
#384 new company license template mmc and mmd

### LINKED ISSUE #384: Create new template to apply for "Company license: MMC  MMM  MDC  MDM  WSL  RIT" and new category "License"
We need to check if we can be using the new "Outcome table" and dynamic listing done by Andrei to add the different licenses connected to a company. And how to select this license when applying for Drug Registration, for example if different licenses can allow applying for the same template. 

The simplest way to create "outcomes" for licenses for this is to add 4 new tables with just at least 2 fields "license number" and document attached related to this.

Purpose of application (radioChoice):

- One of the four option to generate a different type of registration (number for now)
- The import/export would grant permission to apply for "Drug Registrations"

No worries about other templates for this first iteration or adapting current "Drug registration"

---

1. MMC - Modern Medicine
2. MMM
3. MDC - Medical Devices
4. MDM
5. WSL
6. RIT

### PR DESCRIPTION
**Pull request by nmadruga**
Fixes #384

### Linked PRs
Based on PR #400 
Merged changes from branch: PR of #762 [link](https://github.com/openmsupply/application-manager-server/pull/394)- Update to UserRego template

### Changes
- Add new template to apply for "Company license" there will be 2 different templates for that, this first one is basic relates to **Modern medicines** or **Medical devices**. And many fields are only presented with information stored in user's table from UserRego and CompanyRego.
- Documents to be imported as part of this template are reduced from the original list on the Slide 35 (check that only the ones in **bold** which are included in this template): 

1. **LMMD01 – Letter of request (physical submission also required)**
2. LMMD02 – CV of applicant with recent (max 1 year) photo
3. LMMD03 – medical certificate (max 3 months)
4. LMMD04 – Certificate of current residence with photo (max 3 months)
5. LMMD05 – Recent (max 1 year) photo (to be used when printing licence)
6. LMMD06 – Certificate of education level
7. LMMD07 – Criminal record “number 3”
8. LMMD08 – Letter from previous employers documenting at least 3 years professional experience (for private sector must be certified by provincial or capital level) 
9. LMMD09 – Documentation showing resignation from previous employment (for private sector must be certified by provincial or capital level)
10. LMMD10 – Map of location of company
11. LMMD11 - Layout of facilities
12. **LMMD12 – Letter from mother-company authorizing establishment of branch**
13. LMMD13 – Proof of ownership of facilities or contract of rent
14. LMMD14 – Copy of business licence issued by Ministry of Commerce
15. **LMMD15 - Site inspection report by F&D unit**
16. **LMMD16 – Company/manufacturer profile (including site master file and GMP certification where applicable) describing origin of products, staff and qualifications, facilities and equipment, business experience, business development plan**

### CODE REVIEW

File: database/insertData/laos/03_template_license_MMC_MMD.js [hunk 1]
```diff
@@ -0,0 +1,1559 @@
...
+                            {
+                              operator: "objectProperties",
+                              children: [
+                                "currentUser.userId"
+                              ]
+                            },
+                            "user.secondary"
+                          ]
+                        }
+                      }
```
Comment (CarlosNZ) [inline database/insertData/laos/03_template_license_MMC_MMD.js:451]: The reason this isn't showing up is you need to build the response object using the `buildObject` operator, like you did with the others. The stored value for "secondary" is a plain string, but the default response needs to be `{ text: <string> }`

File: database/insertData/laos/03_template_license_MMC_MMD.js [hunk 2]
```diff
@@ -0,0 +1,1559 @@
...
+              ]
+            }
+            templateCategoryToTemplateCategoryId: {
+              connectByCode: { code: "license" }
+            }
+            ${joinFilters}
+            templateActionsUsingId: {
+              create: [
+                ${coreActions}
+                ${devActions}
```
Comment (CarlosNZ) [inline database/insertData/laos/03_template_license_MMC_MMD.js:1393]: Prob remove `devActions`?

File: database/insertData/laos/03_template_license_MMC_MMD.js [hunk 3]
```diff
@@ -0,0 +1,1559 @@
...
+                      children: ["applicationData.responses.Q2LicenseType.text"]
+                    }
+                    expiry: {
+                      operator: "objectProperties"
+                      children: ["31/01/2022"]
+                    }
+                    company_id: {
+                      operator: "objectProperties"
+                      children: ["currentUser.orgId"]
+                    }
```
Comment (CarlosNZ) [inline database/insertData/laos/03_template_license_MMC_MMD.js:1456]: Are you just trying to insert hard-coded values? This isn't working, but you can just put the literal values in:
```
expiry: "31/01/2022",
```

**Review by CarlosNZ — CHANGES_REQUESTED** [review]
Cool, coming along nicely. 

I got mostly "Can't resolve object" inserted into the "license" table for most fields, but we can figure that out once we've locked this workflow down a bit more clearly.

File: database/insertData/laos/03_template_license_MMC_MMD.js (hunk 1 above)
Comment (CarlosNZ) [reply database/insertData/laos/03_template_license_MMC_MMD.js:451]: Although, now that I think about it,  I thought @andreievg had made it so plain strings would be automatically converted to "{text:...}" when evaluating, but that might have just been for shortText. Will need to look into this further (bring it up at meeting this afternoon)

File: database/insertData/laos/03_template_license_MMC_MMD.js [hunk 4]
```diff
@@ -0,0 +1,1559 @@
...
+exports.queries = [
+  `mutation {
+    createTemplate(
+      input: {
+          template: {
+            code: "CompanyLicense1"
+            name: "Company License for Modern medicines or Medical devices"
+            isLinear: false # CHANGE THIS
+            status: AVAILABLE
+            startMessage: "## Apply for a company license for:\\n Modern medicines or Medical devices.\\n\\n**You will be required to upload the following documents:**\\n- LMMD01: Letter of request \\n- LMMD12: Letter from mother-company authorizing establishment of branch\\n- LMMD15: Site inspection report by F&D unit\\n- LMMD16: Company/manufacturer profile (including site master file and GMP certification where applicable)"
```
[the remark this answers is not in the document]
Comment (nmadruga) [reply database/insertData/laos/03_template_license_MMC_MMD.js:18]: Yes. I agree. Changing to **bold** now : )

[the remark this answers is not in the document]
Comment (nmadruga) [reply database/insertData/laos/03_template_license_MMC_MMD.js]: Done: 3720b4d

File: database/insertData/laos/03_template_license_MMC_MMD.js [hunk 5]
```diff
@@ -0,0 +1,1559 @@
...
+                                    ]
+                                  },
+                                  "user.firstName"
+                                ]
+                              }
+                            }
+                          ]
+                        }
+                      }
+                      {
```
[the remark this answers is not in the document]
Comment (nmadruga) [reply database/insertData/laos/03_template_license_MMC_MMD.js:62]: Yes. Sure, I had it like that, but changed I think for consistency

File: database/insertData/laos/03_template_license_MMC_MMD.js [hunk 6]
```diff
@@ -0,0 +1,1559 @@
...
+                        code: "Q12EducationHistory"
+                        index: 190
+                        title: "Education List"
+                        elementTypePluginCode: "listBuilder"
+                        category: QUESTION
+                        isEditable: false
+                        parameters: {
+                          label: "Education history"
+                          displayType: "card"
```
[the remark this answers is not in the document]
Comment (nmadruga) [reply database/insertData/laos/03_template_license_MMC_MMD.js:460]: I changed to `table` so it's the same as in User Rego

File: database/insertData/laos/03_template_license_MMC_MMD.js [hunk 7]
```diff
@@ -0,0 +1,1559 @@
...
+                              {
+                                operator: "objectProperties",
+                                children: [
+                                  "currentUser.userId"
+                                ]
+                              },
+                              "user.universityHistory"
+                            ]
+                          }
+                        }
```
[the remark this answers is not in the document]
Comment (nmadruga) [reply database/insertData/laos/03_template_license_MMC_MMD.js:478]: Let's see : )

[the remark this answers is not in the document]
Comment (nmadruga) [reply database/insertData/laos/03_template_license_MMC_MMD.js]: Hmmm not quite. I think it also uses the field type to display

File: database/insertData/laos/03_template_license_MMC_MMD.js (hunk 1 above)
Comment (nmadruga) [reply database/insertData/laos/03_template_license_MMC_MMD.js:424]: No, as discussed in the meeting. Will need to use the `buildObject`, thanks for looking into that for me : )

File: database/insertData/laos/03_template_license_MMC_MMD.js (hunk 3 above)
Comment (nmadruga) [reply database/insertData/laos/03_template_license_MMC_MMD.js]: Cool. Changing to that now : D

**Comment by nmadruga** [comment]
Ok. Leaving now. Have done the `outcomesDisplay` for licenses, had to remove 2 of the fields since they were still getting "Can't resolve object", but fine for now. We can improve this later... Just not very happy about layout of outcomes : (

Screen Shot 2021-06-10 at 5 29 48 PM

**Comment by nmadruga** [comment]
Merging this one so it's easier for us to work on other fixes if this new template on `master`.

**Pull request merged**